Add enablers for custom/in-app virtual keyboards

jpnurmi opened the pull request:
## Description

This PR introduces a new public class `TextInputSource` used to create `TextInputConnection` instances from within `TextInput.attach()`. This makes it possible to replace the default method channel -based text input source with a custom in-app virtual keyboard on platforms that do not have a system-provided keyboard.

## Related Issues

Closes #68988

## Tests

I added the following tests:

- 'TextInputSource' group in `flutter/test/services/text_input_test.dart`.

## Checklist

Before you create this PR, confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Did any tests fail when you ran them? Please read [Handling breaking changes].

- [x] No, no existing tests failed, so this is *not* a breaking change.
- [ ] Yes, this is a breaking change. *If not, delete the remainder of this section.*

<!-- Links -->
[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

jpnurmi committed:
[WIP] just the old PR rebased so far...

Closes: #68988

jpnurmi commented:
/cc @justinmc 

P.S. The combined diff looks a bit intimidating. It might be better to look at the individual commits, which are in logical steps.

pedromassangocode commented on packages/flutter/lib/src/services/text_input.dart:
```diff
@@ -934,11 +930,128 @@ class TextInputConnection {
+  ///
+  /// After calling this method, the text input control might disappear if no
+  /// other client attaches to it within this animation frame.
+  @mustCallSuper
+  void close() {
+    if (attached) {
+      TextInput._instance._clearClient();
+    }
+    assert(!attached);
+  }
```
Is this supposed to call `super.close()`? I see the @mustCallSuper annotation there.

justinmc commented on packages/flutter/lib/src/services/text_input.dart:
```diff
@@ -934,11 +930,128 @@ class TextInputConnection {
+  ///
+  /// After calling this method, the text input control might disappear if no
+  /// other client attaches to it within this animation frame.
+  @mustCallSuper
```
Is it ok that `close` and `connectionClosedReceived` don't call super because this is an abstract class?

jpnurmi replied:
The reason I marked this as @mustCallSuper is that the private TextInput._clearClient() needs to be called at this point. It schedules hiding the keyboard and clears the current connection.

Alternatively, we could add something like TextInput.detach() and document that the TextInputConnection.close() implementation should call it.

justinmc reviewed (commented):
Just a quick initial review.  Thanks for the clean commit history.

Could you walk me through how these changes are used by your virtual keyboard project?  I see TextInput being used in [keyboard.dart](https://github.com/jpnurmi/virtual_keyboard/blob/master/lib/src/keyboard.dart).  Just want to make sure I understand how this will be used by an in-app keyboard.  We might also want to include an example in the docs, or just some more information, for anyone else that might want to implement something like an in-app keyboard.

CC @gspencergoog for any desktop implications, and tagging @LongCatIsLooong as another reviewer.

jpnurmi commented on packages/flutter/lib/src/services/text_input.dart:
```diff
@@ -862,15 +862,12 @@ abstract class TextInputClient {
 ///    the system's text input using a [TextInputConnection].
-class TextInputConnection {
-  TextInputConnection._(this._client)
+abstract class TextInputConnection {
```
I was thinking that since the constructor was private before, nothing outside text_input.dart could have made TextInputConnection instances. Are there any other implications? Would it be better or safer to have non-functional stub implementations in TextInputConnection?

    void show() {
        assert(attached);
    }

There are a few public API changes, but my understanding is that none of them are breaking:

- TextInputConnection constructor made public
  - Needed for subclassing
- TextInputConnection gained three new methods: hide(), setClient(), clearClient()
  - The respective methods already exist in the "TextInput" method channel API
  - The method channel calls were previously made directly from TextInput

jpnurmi reviewed (commented):
> Just a quick initial review.

Thanks!

> Could you walk me through how these changes are used by your virtual keyboard project?

Oh, it's not mine, actually. I just found someone's old (unmaintained?) package on pub.dev and used it as a testbed. It would be awesome to port LatinIME from AOSP to Flutter, but I have a feeling it's not as straight-forward as it sounds. :)

> I see TextInput being used in keyboard.dart. Just want to make sure I understand how this will be used by an in-app keyboard.

Framework => VKB:

  - EditableText calls TextInput.attach() to create a connection, and passes itself as a "client"
  - EditableText calls TextInputConnection methods to communicate with the keyboard
  - TextInputConnection handles text input related requests from the framework
    - when to show or hide the keyboard
    - what is the current input configuration/type (text, multiline, numeric, ...)
    - the default implementation relays these over the method channel to the platform plugin
    - a custom implementation can communicate with a local keyboard widget

VKB => Framework:

  - Keyboard calls TextInputClient methods to communicate with the framework
    - update the editing value
    - perform an action
    - ...

The connection factory is the glue between. It gives access to the current TextInputClient, and allows returning a custom TextInputConnection. This way a keyboard can have full control over both ways of the text input communication.

> We might also want to include an example in the docs, or just some more information, for anyone else that might want to implement something like an in-app keyboard.

Yeah, I'd imagine covering this topic thoroughly with examples and all that could easily result in a separate dedicated doc page.

justinmc commented:
Thank you for the explanation!  Are you actually communicating over the TextInputChannel in [your forked virtual keyboard project](https://github.com/jpnurmi/virtual_keyboard), or are you using that project in another app?  Could you point me to where the usage of the changes in this PR happens?

I think you're right that this shouldn't be a breaking change.

jpnurmi commented:
Yeah, the experiment is in the [text-input-connection-factory](https://github.com/jpnurmi/virtual_keyboard/commits/text-input-connection-factory/) branch.

Registration:
- `_VirtualKeyboardState.initState()` calls `TextInput.registerConnectionFactory()`:
  https://github.com/jpnurmi/virtual_keyboard/blob/text-input-connection-factory/lib/src/keyboard.dart#L184

Communication FW => VKB:
- `class _VirtualKeyboardConnection extends TextInputConnection`:
  https://github.com/jpnurmi/virtual_keyboard/blob/text-input-connection-factory/lib/src/keyboard.dart

Communication VKB => FW:
- `_VirtualKeyboardState._updateEditingState()` calls `TextInputClient.updateEditingValue()` to update `EditableText` content:
  https://github.com/jpnurmi/virtual_keyboard/blob/text-input-connection-factory/lib/src/keyboard.dart#L288

- `_VirtualKeyboardState._performAction()` calls `TextInputClient.performAction()` to perform newline/done/go etc. actions:
  https://github.com/jpnurmi/virtual_keyboard/blob/text-input-connection-factory/lib/src/keyboard.dart#L294

The VKB is using a [helper model](https://github.com/jpnurmi/virtual_keyboard/blob/master/lib/src/model.dart) to manage the text input state. It's a Dart port of the [TextInputModel](https://github.com/flutter/engine/blob/master/shell/platform/common/cpp/text_input_model.h) used by the Windows & Linux platform shells.

LongCatIsLooong commented on packages/flutter/lib/src/services/text_input.dart:
```diff
@@ -1184,123 +1363,33 @@ class TextInput {
+  ///
+  /// By default, [TextInputConnection] communicates with the platform's text
+  /// input plugin via [SystemChannels.textInput].
+  static void registerConnectionFactory(TextInputConnectionFactory? factory) {
```
What would become of the currently connected client if a different factory is registered? Should we make this thing an `InheritedWidget` so the current client (that has an open connection) can switch to the new factory, or we shouldn't support that use case at all(i.e., this should only be called before `runApp` and should never be called again)?

jpnurmi replied:
It's fine to register and unregister a factory on the fly, but it's a factory so it will only affect the next client/connection.

jpnurmi replied:
> nit: can this be static void registerConnectionFactory(TextInputConnectionFactory factory = _defaultConnectionFactory) {?

I wasn't able to resolve this request, because `TextInputConnection` (it increments a running client ID) doesn't have a const constructor, so I couldn't make `registerConnectionFactory` const either.

LongCatIsLooong replied:
> It's fine to register and unregister a factory on the fly, but it's a factory so it will only affect the next client/connection.

This means that if there's an active `TextField` and the factory is changed on the fly, then the user won't be able to use the new text input device unless they switch to a different field and switch back right? IMO we should either document that the developer shouldn't change the factory on the fly unless they can guarantee there's no active connection, or provide a means to notify the current client that the factory has been changed and it needs to reestablish the connection.

jpnurmi replied:
I'm feeling a bit unsure if it's a good idea to make it fully automatic, but I have added a check for an active connection in `setInputConnectionFactory()`. It hides the old text input control and re-attaches the same client so that a new connection type is established.

LongCatIsLooong commented on packages/flutter/lib/src/services/text_input.dart:
```diff
@@ -1126,7 +1332,6 @@ RawFloatingCursorPoint _toTextPoint(FloatingCursorDragState state, Map<String, d
 class TextInput {
   TextInput._() {
     _channel = SystemChannels.textInput;
```
Does it still make sense for the `TextInput` to hold the `SystemChannel`?

jpnurmi replied:
There were a couple of reasons I left it here.
- `TextInput.setChannel()` for testing
- I'm not sure what to do with `TextInput.finishAutofillContext()`, which is calling the method channel directly and bypassing any text input connection. I was tempted to add `TextInputConnection.finishAutofillContext()`, but it would change the behavior so that it would only do something if a client is attached. Any idea what is the desired behavior? Apparently `TextInput.finishAutofillContext` is not implemented on Linux, at least.

LongCatIsLooong commented on packages/flutter/lib/src/services/text_input.dart:
```diff
@@ -934,11 +930,128 @@ class TextInputConnection {
+class _TextInputChannelConnection extends TextInputConnection {
+  _TextInputChannelConnection(TextInputClient client, this._channel)
+      : super(client) {
+    _channel.setMethodCallHandler(_handleTextInputInvocation);
```
Oh there's no `attach` in `TextInputConnection`. I think we might want to expose more lifecycle methods (`onAttach`, `onDetach`, etc.).

jpnurmi replied:
Thanks for the proposal. I added `onAttach` and `onDetach`.

jpnurmi commented on packages/flutter/lib/src/services/text_input.dart:
```diff
@@ -893,10 +890,10 @@ class TextInputConnection {
+  void show();
+
+  /// Requests that the text input control is hid.
+  void hide();
```
- I have added an explanation that this method is called by the framework when the text input control should hide. I added also see also links to each other in hide() and close().
- Cool, I hadn't realized that @protected can be called from anywhere in the same file.

LongCatIsLooong reviewed (commented):
Sorry for the delay. LGTM in general, left a few comments regarding the lifecycle.

jpnurmi commented:
The PR is slowly growing, but I like where we're heading. ;) With the latest version, it's now possible to switch the text input control/factory on the fly.

![vkb](https://user-images.githubusercontent.com/140617/98748886-e2bf2900-23ba-11eb-95f6-0ce4512028cf.gif)

jpnurmi commented on packages/flutter/lib/src/services/text_input.dart:
```diff
@@ -934,11 +976,132 @@ class TextInputConnection {
+    required TextAlign textAlign,
+  });
+
+  /// Use [TextInput.detach] instead.
+  @Deprecated(
+    'Use TextInput.detach instead. '
+    'This feature was deprecated after v1.24.0-7.0.pre.'
+  )
+  void close() {
+    TextInput.detach(_client);
+    assert(!attached);
+  }
+
+  /// Use [TextInput.reset] instead.
+  @Deprecated(
+    'Use TextInput.reset instead. '
+    'This feature was deprecated after v1.24.0-7.0.pre.'
+  )
+  void connectionClosedReceived() {
+    TextInput.reset();
+    assert(!attached);
+  }
```
What do you think about deprecating these and providing the respective functionality in TextInput instead? This change solves the problem that these had to be marked as @mustCallSuper earlier because they were calling private API in TextInput.

jpnurmi commented on packages/flutter/lib/src/services/text_input.dart:
```diff
@@ -1311,59 +1455,45 @@ class TextInput {
+  /// This function should be called to reset the current text input connection
+  /// in case the platform sent a notification informing the connection is
+  /// closed.
+  static void reset() {
```
Unsure about the naming here as well

jpnurmi commented on packages/flutter/lib/src/services/text_input.dart:
```diff
@@ -852,25 +856,43 @@ abstract class TextInputClient {
+  /// Framework notified of a restored connection.
+  ///
+  /// [TextInputClient] should show the text input control if appropriate.
+  void connectionRestored();
```
Unsure about the naming here. The client needs to be notified in a way or another when a connection is replaced on the fly.

jpnurmi commented:
Would you prefer to offload some of the preparational steps to a separate PR?

LongCatIsLooong commented on packages/flutter/lib/src/services/text_input.dart:
```diff
@@ -934,11 +976,132 @@ class TextInputConnection {
+
+  @override
+  void onDetach() {
+    _channel.setMethodCallHandler(null);
```
Maybe we need an empty handler that simply ditches messages? From the `_DefaultBinaryMessenger` code, the unhandled messages will be pushed into a buffer: https://github.com/flutter/flutter/blob/8436c6ab7e80ac545c6ea7125578b96724962faf/packages/flutter/lib/src/services/binding.dart#L274-L301

(I'm not entirely sure)

jpnurmi replied:
I've made the method call handler static and let it remain to avoid leaking unhandled messages. There's a check to see if the current connection is a method channel -based connection and the incoming messages are simply ignored with custom sources.

LongCatIsLooong reviewed (commented):
So these are the lifecycle events of an `InputConnection` (excluding deprecated ones): `onAttach` -> `setClient` -> `clearClient` -> `onDetach`?

jpnurmi commented:
The branch was getting messy and falling behind so I've rebased and cleaned up the commits. Also, I felt like a factory method was no longer appropriate for what we were doing with it. I liked the term "input source" that was used in some comments, so I turned the factory method into a `TextInputSource` class. What do you think about this approach?

jpnurmi commented on packages/flutter/lib/src/services/text_input.dart:
```diff
@@ -852,25 +856,43 @@ abstract class TextInputClient {
+/// over the [SystemChannels.textInput] method channel. See [SystemChannels.textInput]
+/// for more details about the method channel messages.
+///
+/// The default implementation can be overridden by registering a factory method
```
I must say I'm struggling a bit with the docs. I'm feeling quite unsure where to mention and what, and there's some overlap in various class/method descriptions here and there...

LongCatIsLooong replied:
Dartdoc supports macros: https://stackoverflow.com/questions/62763658/how-do-i-create-dartdoc-macros-templates-in-dart-and-flutter

LongCatIsLooong commented on packages/flutter/lib/src/services/text_input.dart:
```diff
@@ -1142,38 +1374,34 @@ class TextInput {
-    TextInputAction.previous,
-    TextInputAction.newline,
-  ];
+  /// Sets the [TextInputSource] used to attach and detach text input clients.
```
this sounds slightly confusing. "Sets the [TextInputSource] for [TextInputClient]s to attach to, when [attach] is called"?

LongCatIsLooong commented on packages/flutter/lib/src/services/text_input.dart:
```diff
@@ -1142,38 +1374,34 @@ class TextInput {
-    TextInputAction.newline,
-  ];
+  /// Sets the [TextInputSource] used to attach and detach text input clients.
+  /// The text input source is responsible for creating [TextInputConnection]
```
nit: I think it would be nice to have an overview of this somewhere. Maybe in the [TextInput] documentation?

LongCatIsLooong commented on packages/flutter/lib/src/services/text_input.dart:
```diff
@@ -1183,128 +1411,41 @@ class TextInput {
 
-  late MethodChannel _channel;
+  TextInputSource? _customSource;
+  TextInputSource get _currentSource => _customSource ?? defaultSource;
```
Why not this?
```suggestion
  TextInputSource _currentSource = defaultSource;
```

LongCatIsLooong commented on packages/flutter/lib/src/services/text_input.dart:
```diff
@@ -1425,3 +1552,55 @@ class TextInput {
+
+/// An interface for attaching and detaching [TextInputClient].
+///
+/// [TextInputSource] creates is responsible for creating [TextInputConnection]
```
typo

LongCatIsLooong commented on packages/flutter/lib/src/services/text_input.dart:
```diff
@@ -1425,3 +1552,55 @@ class TextInput {
+/// creates [TextInputConnection] instances that communicate with the platform
+/// text input plugin over the [SystemChannels.textInput] method channel. See
+/// [SystemChannels.textInput] for more details about the text input method
+/// channel messages.
```
super nit: I feel we can combine this paragraph and the dedicated "Custom text input sources" section since they're both example usages?

LongCatIsLooong commented on packages/flutter/lib/src/services/text_input.dart:
```diff
@@ -856,25 +856,51 @@ abstract class TextInputClient {
+///
+/// ### Connect to a different input source
+///
+/// The default [TextInputConnection] can be replaced by setting a custom
```
nit: this sounds slightly inaccurate, `setSource` doesn't "replace" the connection, but the developer may decide to replace the connection in `didUpdateInputSource`, is that correct?

LongCatIsLooong reviewed (commented):
Sorry for the delay. It looks great, thank you!  This is a big change and may affect web, so /cc @justinmc @nturgut

jpnurmi commented:
It's good to hear that things are starting to look good. I'm starting to feel that this is getting a bit out of hand, though. To make it easier to focus on the important parts, I would propose to start with the preparatory `TextInputConnection` vs. `_TextInputChannelConnection` changes: https://github.com/flutter/flutter/pull/72803. We can get back to the `TextInputSource` changes here once the foundation has been laid.

jpnurmi committed:
Merge remote-tracking branch 'origin/master' into text-input

jpnurmi committed:
Resolved a few review comments

jpnurmi committed:
Add TextInputSource docs & example

jpnurmi commented:
I have revised the docs and added a dummy VKB example snippet. I left out some irrelevant parts to keep it as short as possible. A full runnable version here: https://gist.github.com/jpnurmi/c70eb5226a26db9aa534f07ae58aa8c6

Some more elaborate examples for inspiration:
- https://github.com/jpnurmi/flutter_vkb_example
- https://github.com/jpnurmi/flutter_keypad_example

jpnurmi committed:
Fix typo and align phrasing

jpnurmi committed:
Update AUTHORS

justinmc reviewed (approved):
LGTM 👍 

Thanks for taking the time to write all of the docs and including that big example, it should be really helpful.

Let me make one more effort to see if someone from web wants to look at this, otherwise I think this is good to go.

yjbanov commented:
/cc @mdebbar

mdebbar commented:
Thanks for the contribution, @jpnurmi!

The changes look good to me. One thing to keep in mind though, is that on web, the framework doesn't handle the keyboard events directly. Instead, we let the browser handle them, and we get `change` events from the browser's input element. See [here](https://github.com/flutter/flutter/blob/dcc4fdd5588d9886953fae7b694f31e465a01916/packages/flutter/lib/src/rendering/editable.dart#L514-L517) to find how we ignore key events on the web.

I'm not sure how that would work with custom input sources. I'm assuming we have to allow the framework to handle key events in that case, right?

mdebbar reviewed (commented):
Thanks for the contribution, @jpnurmi!

The changes look good to me. One thing to keep in mind though, is that on web, the framework doesn't handle the keyboard events directly. Instead, we let the browser handle them, and we get `change` events from the browser's input element. See [here](https://github.com/flutter/flutter/blob/dcc4fdd5588d9886953fae7b694f31e465a01916/packages/flutter/lib/src/rendering/editable.dart#L514-L517) to find how we ignore key events on the web.

I'm not sure how that would work with custom input sources. I'm assuming we have to allow the framework to handle key events in that case, right?

justinmc commented:
@jpnurmi Could you try out your virtual keyboard example in Flutter web and see what happens?

jpnurmi commented:
Thanks for the comments. I'm glad you brought this up.

A custom VKB works fine on any platform because it basically communicates directly with `EditableText`, but it does conflict with physical keyboards. Any keys that are processed directly (movement/selection, delete...) continue working, but it seems not to be possible to type in text with a physical keyboard after overriding the text input source. This is simply because with a custom `TextInputSource`, messages from `SystemChannels.textInput` end up being discarded. :-1:

After digging a bit deeper, I can see now that the Linux platform plugin maintains its internal `TextInputModel` under the hood. The model is filled up with text from incoming native key events. With a custom text input source, the framework never sends any editing state changes back to the platform plugin, so the internal model keeps piling up with bogus text. Ouch, back to the drawing board...

jpnurmi commented:
I have a new proposal that I'd like someone to double-check when you have a few spare minutes: https://github.com/jpnurmi/flutter/commit/62aebc3aff21a3c6f1fd2398e330d26c40b9b5f9

In a nutshell, we could categorize text input calls to state changes vs. visual changes. When a custom input control is registered, we clearly cannot cut off the text input method channel, but state changes must continue to be delivered to the platform backend. This ensures that physical keyboards managed by the platform plugin and custom virtual keyboards continue to work seamlessly together. Visual changes, however, would only be communicated to the current text input control to make sure only one input control shows up. What do you think?

justinmc commented:
@jpnurmi I like that proposal.  So your virtual keyboard would implement TextInputHandler and register itself via setInputControl?

What about unregistering the current _customControl?

Also, would there ever be a reason to register multiple _customControls?  It seems like all of this could be more generic to support things like that, but there's probably no reason to do that if there is always one _platformControl and optionally one _customControl.

justinmc commented:
@jpnurmi Are you still available to work on this PR?

jpnurmi commented:
@justinmc Yes, I'm still very much interested in having something like this in Flutter! Should we continue exploring the "TextInputControl" proposal?

> @jpnurmi I like that proposal. So your virtual keyboard would implement TextInputHandler and register itself via setInputControl?

s/TextInputHandler/TextInputControl/ but basically yes :)

> What about unregistering the current _customControl?
>
> Also, would there ever be a reason to register multiple _customControls? It seems like all of this could be more generic to support things like that, but there's probably no reason to do that if there is always one _platformControl and optionally one _customControl.

I think we could allow registering multiple input controls, but perhaps only activate one at a time? The currently active input control would be the one that shows up, but the others would also get to process text input events.

jpnurmi commented:
Here's a documented version of `TextInputHandler` and `TextInputControl`. Only one (visual) text input control can be set at a time, but it's now possible to install multiple (non-visual) text input handlers.

https://github.com/flutter/flutter/compare/696e4f70d1a7a9b9b793aa3fc4ad2d8d3d8eac17...jpnurmi:text-input-handler?expand=1

Updated examples:
- https://github.com/jpnurmi/flutter_minimal_vkb_example
- https://github.com/jpnurmi/flutter_vkb_example
- https://github.com/jpnurmi/flutter_keypad_example

The old tests still need to be updated. I'll abandon this PR and leave it as a reference, and open a separate one for the new implementation when it's finished.

jpnurmi commented:
A new attempt: #76072
